In [1]:
import pandas as pd
import numpy as np

In [2]:
orders = pd.read_csv(r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_orders_dataset.csv")
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [3]:
print(orders.shape)
print(orders.columns)



(99441, 8)
Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')


In [4]:
print(orders.dtypes)


order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object


In [5]:
print(orders.isnull().sum())

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


In [6]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])
orders["order_approved_at"] = pd.to_datetime(orders["order_approved_at"])
orders["order_delivered_carrier_date"] = pd.to_datetime(orders["order_delivered_carrier_date"])
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])
orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])


In [7]:
orders["order_status"].value_counts(normalize=True)


order_status
delivered      0.970203
shipped        0.011132
canceled       0.006285
unavailable    0.006124
invoiced       0.003158
processing     0.003027
created        0.000050
approved       0.000020
Name: proportion, dtype: float64

In [8]:
orders["order_purchase_timestamp"].min()

Timestamp('2016-09-04 21:15:19')

In [9]:
orders["order_purchase_timestamp"].max()

Timestamp('2018-10-17 17:30:18')

In [10]:
orders["delivery_days"] = orders["order_delivered_customer_date"]-orders["order_purchase_timestamp"]

In [11]:
orders["delivery_days"]

0        8 days 10:28:40
1       13 days 18:46:08
2        9 days 09:27:40
3       13 days 05:00:36
4        2 days 20:58:23
              ...       
99436    8 days 05:13:56
99437   22 days 04:38:58
99438   24 days 20:37:34
99439   17 days 02:04:27
99440    7 days 16:11:00
Name: delivery_days, Length: 99441, dtype: timedelta64[us]

In [12]:
#1 ans
len(orders[orders["order_status"] =="delivered"])


96478

In [13]:
orders.shape

(99441, 9)

In [14]:
#3 ans
delivered = len(orders[orders["order_status"] =="delivered"])
total_orders = len(orders)
print(total_orders)
print("Deliverd Orders in % :- ",(delivered / total_orders) *100)

99441
Deliverd Orders in % :-  97.02034372140264


In [15]:
orders["orders_months"] = orders["order_purchase_timestamp"].dt.to_period("M")
orders["orders_months"].head()

0    2017-10
1    2018-07
2    2018-08
3    2017-11
4    2018-02
Name: orders_months, dtype: period[M]

In [16]:
orders["orders_months"].sort_values()

4541     2016-09
10071    2016-09
4396     2016-09
30710    2016-09
10381    2016-10
          ...   
1801     2018-09
88500    2018-10
31891    2018-10
68373    2018-10
60938    2018-10
Name: orders_months, Length: 99441, dtype: period[M]

In [17]:
orders["delivery_days"].mean()

Timedelta('12 days 13:24:31.879068')

In [18]:
orders["delivery_days"].min()

Timedelta('0 days 12:48:07')

In [19]:
orders["delivery_days"].max()


Timedelta('209 days 15:05:12')

In [20]:

orders["Delayed_orders"] = (orders["order_delivered_customer_date"] > orders["order_estimated_delivery_date"])

In [21]:
orders["delay_days"]=(orders["order_delivered_customer_date"]-orders["order_estimated_delivery_date"]).dt.days
print(orders["delay_days"])

0        -8.0
1        -6.0
2       -18.0
3       -13.0
4       -10.0
         ... 
99436   -11.0
99437    -2.0
99438    -6.0
99439   -21.0
99440   -18.0
Name: delay_days, Length: 99441, dtype: float64


In [22]:
orders["delay_days"].mean()

np.float64(-11.876881296902857)

In [23]:
orders[orders["Delayed_orders"] == True]["delay_days"].mean()

np.float64(8.86725437587837)

# Business Insight 


In [24]:
# Total orders to delivered order is 97.020% with percentage
# Faser delivery all time is 0 days 12:48:07
# Sowest delivery all time to delivery orders  209 days 15:05:12
# Delayed orders average to delay day : - 8.87 days

## Check the revenue payments and sales for the trend according 


In [25]:
order_items = pd.read_csv(r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_order_items_dataset.csv")
payments = pd.read_csv(r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_order_payments_dataset.csv")

In [26]:
order_items.shape


(112650, 7)

In [27]:
order_items.columns


Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [28]:
# Q1. Total product sales value
order_items["price"].sum()

np.float64(13591643.7)

In [29]:
# Q2. Total freight revenue/cost
order_items["freight_value"].sum()
#  all order to placed by company is 2251909

np.float64(2251909.54)

In [30]:

payments.shape

(103886, 5)

In [31]:
payments.columns


Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')

In [32]:
# Q3. Total payment value
payments["payment_value"].sum()
#  All total payments for to sale the product is :  16008872

np.float64(16008872.120000001)

In [33]:
#  Q4. Average product price
order_items["price"].mean()


np.float64(120.65373901464713)

In [34]:
# Q5. Most expensive product item
order_items["price"].max()
#  big amount for sale by company is 13664 

np.float64(6735.0)

In [35]:
# Q6. Cheapest product item
order_items["price"].min()
#  chepest price for the product the comany sale is 0 

np.float64(0.85)

In [36]:
# Q7. Payment method analysis
payments["payment_type"].value_counts()
# there are customer aretype of  payments_type during the purchasing product     credit_card    76795
#                                                                                boleto         19784
#                                                                                voucher         5775
#                                                                                debit_card      1529
#                                                                                not_defined        3

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [37]:
# Q8. Payment method percentage
payments["payment_type"].value_counts(normalize=True) * 100

payment_type
credit_card    73.922376
boleto         19.043952
voucher         5.558978
debit_card      1.471806
not_defined     0.002888
Name: proportion, dtype: float64

In [38]:
# Q9 — Installment analysis
payments["payment_installments"].min()


np.int64(0)

In [39]:
payments["payment_installments"].max()


np.int64(24)

In [40]:

payments["payment_installments"].mean()


np.float64(2.853348863176944)

In [41]:
products = pd.read_csv(r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_products_dataset.csv")

In [42]:
products.shape

(32951, 9)

In [43]:
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

In [44]:
# 1️⃣ Product Categories
products["product_category_name"].nunique
#  73
# business insight 
#  there are check a how many products name in table is 73 

<bound method IndexOpsMixin.nunique of 0                               perfumaria
1                                    artes
2                            esporte_lazer
3                                    bebes
4                    utilidades_domesticas
                       ...                
32946                     moveis_decoracao
32947    construcao_ferramentas_iluminacao
32948                      cama_mesa_banho
32949               informatica_acessorios
32950                      cama_mesa_banho
Name: product_category_name, Length: 32951, dtype: str>

In [45]:
#  2️⃣ Missing Categories
products["product_category_name"].isnull().sum()
#  610 
# business insight
#  that are 610 rows are affected to not complete data fill that

np.int64(610)

In [46]:
# 3️⃣ Category Distribution
products["product_category_name"].value_counts().head(10)

# product_category_name
# cama_mesa_banho           3029
# esporte_lazer             2867
# moveis_decoracao          2657
# beleza_saude              2444
# utilidades_domesticas     2335
# automotivo                1900
# informatica_acessorios    1639
# brinquedos                1411
# relogios_presentes        1329
# telefonia                 1134
# Name: count, dtype: int64

#  business insight 
#  product categorey wise top 10 most selling  product name 

product_category_name
cama_mesa_banho           3029
esporte_lazer             2867
moveis_decoracao          2657
beleza_saude              2444
utilidades_domesticas     2335
automotivo                1900
informatica_acessorios    1639
brinquedos                1411
relogios_presentes        1329
telefonia                 1134
Name: count, dtype: int64

In [47]:
# 4️⃣ Product Photos
products["product_photos_qty"].mean()
# np.float64(2.1889861166939797)

#  business insight 
#  there are most of the product average  pic in the table is 2 

np.float64(2.1889861166939797)

In [48]:
# 5️⃣ Most Photos
products["product_photos_qty"].max()
# np.float64(20.0)

# business insight 
#  maximum images for one product in the table appoximently 20 

np.float64(20.0)

In [49]:
#  6️⃣ Product Weight
products["product_weight_g"].mean()
# np.float64(2276.4724877841513)

#  business insight 
# there are average weight of product 2.3kg

np.float64(2276.4724877841513)

In [50]:
# 7️⃣ Heaviest Product
products["product_weight_g"].max()
#  np.float64(40425.0)

#  business insight 
#  there are maximum weight for product is 4kg

np.float64(40425.0)

In [51]:
#  8️⃣ Product Dimensions
products[["product_height_cm",
         "product_length_cm",
         "product_width_cm"]].mean()

# product_height_cm    16.937661
# product_length_cm    30.815078
# product_width_cm     23.196728
# dtype: float64

#  business insight 
#  there are check the average product size to  calulate in cm

product_height_cm    16.937661
product_length_cm    30.815078
product_width_cm     23.196728
dtype: float64

In [52]:
# 9️⃣ Missing Data Check
products.isnull().sum()

# product_id                      0
# product_category_name         610
# product_name_lenght           610
# product_description_lenght    610
# product_photos_qty            610
# product_weight_g                2
# product_length_cm               2
# product_height_cm               2
# product_width_cm                2
# dtype: int64

#  business insight 
#  column wise to check a all empty value in table 

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

## Product + Sales Analysis


In [53]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [54]:
# Q1. Total Products Sold
order_items.groupby("product_id").size().reset_index(name = "total_itms_sold")

,product_id,total_itms_sold
0,00066f42aeeb9f3007548bb9d3f33c38,1
1,00088930e925c41fd95ebfe695fd2655,1
2,0009406fd7479715e4bef61dd91f2462,1
3,000b8f95fcb9e0096488278317764d19,2
4,000d9be29b5207b54e86aa1b1ac54872,1
...,...,...
32946,fff6177642830a9a94a0f2cba5e476d1,2
32947,fff81cc3158d2725c0655ab9ba0f712c,1
32948,fff9553ac224cec9d15d49f5a263411f,1
32949,fffdb2d0ec8d6a61f0a0a0db3f25b441,5


In [55]:
#  Q2. Top 10 Most-Sold Product
product_counts = order_items.groupby("product_id")["order_item_id"].count()

sorted_counts = product_counts.sort_values(ascending = False).head(10)
sorted_counts

product_id
aca2eb7d00ea1a7b8ebd4e68314663af    527
99a4788cb24856965c36a24e339b6058    488
422879e10f46682990de24d770e7f83d    484
389d119b48cf3043d311335e499d9c6b    392
368c6c730842d78016ad823897a372db    388
53759a2ecddad2bb87a079a1f1519f73    373
d1c427060a0f73f6b889a5c7c61f2ac4    343
53b36df67ebb7c41585e8d54d6772e08    323
154e7e31ebfa092203795c972e5804a6    281
3dd2a17168ec895c781a9191c1e95ad7    274
Name: order_item_id, dtype: int64

In [56]:
# Q3. Product-wise Total Revenue
sales_revenue = order_items.groupby("product_id")["price"].sum()
print(sales_revenue)

product_id
00066f42aeeb9f3007548bb9d3f33c38    101.65
00088930e925c41fd95ebfe695fd2655    129.90
0009406fd7479715e4bef61dd91f2462    229.00
000b8f95fcb9e0096488278317764d19    117.80
000d9be29b5207b54e86aa1b1ac54872    199.00
                                     ...  
fff6177642830a9a94a0f2cba5e476d1    229.98
fff81cc3158d2725c0655ab9ba0f712c     90.00
fff9553ac224cec9d15d49f5a263411f     32.00
fffdb2d0ec8d6a61f0a0a0db3f25b441    169.95
fffe9eeff12fcbd74a2f2b007dde0c58    249.99
Name: price, Length: 32951, dtype: float64


In [57]:
#  Q4. Top 10 Products by Revenue
sales_revenue.sort_values(ascending= False).head(10)

product_id
bb50f2e236e5eea0100680137654686c    63885.00
6cdd53843498f92890544667809f1595    54730.20
d6160fb7873f184099d9bc95e30376af    48899.34
d1c427060a0f73f6b889a5c7c61f2ac4    47214.51
99a4788cb24856965c36a24e339b6058    43025.56
3dd2a17168ec895c781a9191c1e95ad7    41082.60
25c38557cf793876c5abdd5931f922db    38907.32
5f504b3a1c75b73d6151be81eb05bdc9    37733.90
53b36df67ebb7c41585e8d54d6772e08    37683.42
aca2eb7d00ea1a7b8ebd4e68314663af    37608.90
Name: price, dtype: float64

In [58]:
# Q5. Category-wise Items Sold
product_sales = products.merge(
    order_items,
    on = "product_id"
)
product_sales.groupby("product_category_name")["order_item_id"].count()


product_category_name
agro_industria_e_comercio     212
alimentos                     510
alimentos_bebidas             278
artes                         209
artes_e_artesanato             24
                             ... 
sinalizacao_e_seguranca       199
tablets_impressao_imagem       83
telefonia                    4545
telefonia_fixa                264
utilidades_domesticas        6964
Name: order_item_id, Length: 73, dtype: int64

In [59]:
# Q6. Category-wise Revenue

Revenue_Categorey = product_sales.groupby("product_category_name")["price"].sum()

In [60]:
# Q7. Top 10 Categories by Revenue
Revenue_Categorey.sort_values(ascending=False).head(10)

product_category_name
beleza_saude              1258681.34
relogios_presentes        1205005.68
cama_mesa_banho           1036988.68
esporte_lazer              988048.97
informatica_acessorios     911954.32
moveis_decoracao           729762.49
cool_stuff                 635290.85
utilidades_domesticas      632248.66
automotivo                 592720.11
ferramentas_jardim         485256.46
Name: price, dtype: float64

In [61]:
# Q8. Top 10 Products by Quantity Sold
product_sales.groupby("product_id")["order_item_id"].count().sort_values(ascending=False).head(10)

product_id
aca2eb7d00ea1a7b8ebd4e68314663af    527
99a4788cb24856965c36a24e339b6058    488
422879e10f46682990de24d770e7f83d    484
389d119b48cf3043d311335e499d9c6b    392
368c6c730842d78016ad823897a372db    388
53759a2ecddad2bb87a079a1f1519f73    373
d1c427060a0f73f6b889a5c7c61f2ac4    343
53b36df67ebb7c41585e8d54d6772e08    323
154e7e31ebfa092203795c972e5804a6    281
3dd2a17168ec895c781a9191c1e95ad7    274
Name: order_item_id, dtype: int64

In [62]:
# # Q9. Find the top 10 products by total revenue.
product_sales.groupby("product_id")["price"].sum().sort_values(ascending = False).head(10)

product_id
bb50f2e236e5eea0100680137654686c    63885.00
6cdd53843498f92890544667809f1595    54730.20
d6160fb7873f184099d9bc95e30376af    48899.34
d1c427060a0f73f6b889a5c7c61f2ac4    47214.51
99a4788cb24856965c36a24e339b6058    43025.56
3dd2a17168ec895c781a9191c1e95ad7    41082.60
25c38557cf793876c5abdd5931f922db    38907.32
5f504b3a1c75b73d6151be81eb05bdc9    37733.90
53b36df67ebb7c41585e8d54d6772e08    37683.42
aca2eb7d00ea1a7b8ebd4e68314663af    37608.90
Name: price, dtype: float64

In [63]:
# Q10. Find the average price of products in each category.
product_sales.groupby("product_category_name")["price"].mean()

product_category_name
agro_industria_e_comercio    342.124858
alimentos                     57.634137
alimentos_bebidas             54.602446
artes                        115.802105
artes_e_artesanato            75.583750
                                ...    
sinalizacao_e_seguranca      108.086583
tablets_impressao_imagem      90.703735
telefonia                     71.213978
telefonia_fixa               225.693182
utilidades_domesticas         90.788148
Name: price, Length: 73, dtype: float64

In [64]:
# Q11. Find the top 10 sellers by total revenue.
product_sales.groupby("seller_id")["price"].sum().sort_values(ascending= False).head(10)

seller_id
4869f7a5dfa277a7dca6462dcf3b52b2    229472.63
53243585a1d6dc2643021fd1853d8905    222776.05
4a3ca9315b744ce9f8e9374361493884    200472.92
fa1c13f2614d7b5c4749cbc52fecda94    194042.03
7c67e1448b00f6e969d365cea6b010ab    187923.89
7e93a43ef30c4f03f38b393420bc753a    176431.87
da8622b14eb17ae2831f4ac5b9dab84a    160236.57
7a67c85e85bb2ce8582c35f2203ad736    141745.53
1025f0e2d44d7041d6cf58b6550e0bfa    138968.55
955fee9216a65b617aa5c0531780ce60    135171.70
Name: price, dtype: float64

In [65]:
# Q12. Find the top 10 sellers by number of items sold.
product_sales.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'order_id', 'order_item_id', 'seller_id', 'shipping_limit_date',
       'price', 'freight_value'],
      dtype='str')

In [66]:
product_sales.groupby("seller_id")["order_item_id"].count().sort_values(ascending = False).head(10)

seller_id
6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
955fee9216a65b617aa5c0531780ce60    1499
1025f0e2d44d7041d6cf58b6550e0bfa    1428
7c67e1448b00f6e969d365cea6b010ab    1364
ea8482cd71df3c1969d7b9473ff13abc    1203
7a67c85e85bb2ce8582c35f2203ad736    1171
Name: order_item_id, dtype: int64

In [67]:
# Q13. Find the average product price for each seller.
product_sales.groupby("seller_id")["price"].mean()

seller_id
0015a82c2db000af6aaaf3ae2ecb0532    895.000000
001cca7ae9ae17fb1caed9dfb1094831    104.937364
001e6ad469a905060d959994f1b41e4f    250.000000
002100f778ceb8431b7a1020ff7ab48f     22.445455
003554e2dce176b5555353e4f3555ac8    120.000000
                                       ...    
ffcfefa19b08742c5d315f2791395ee5     69.900000
ffdd9f82b9a447f6f8d4b91554cc7dd3    105.060000
ffeee66ac5d5a62fe688b9d26f83f534    131.418571
fffd5413c0700ac820c7069d66d98c89    148.562295
ffff564a4f9085cd26170f4732393726     71.315000
Name: price, Length: 3095, dtype: float64

In [68]:
#  radhe radhe 


In [69]:
#  Q2 — Unsold Products
products.columns

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='str')

In [70]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [71]:
unsold_check = products.merge(
    order_items,
    on="product_id",
    how="left"
)
unsold_check["order_item_id"].isnull().sum()

np.int64(0)

In [72]:

unsold_check["order_item_id"].isnull().sum()

np.int64(0)

In [73]:
unsold_products = unsold_check[
    unsold_check["order_item_id"].isnull()
]

In [74]:
print(unsold_products)

Empty DataFrame
Columns: [product_id, product_category_name, product_name_lenght, product_description_lenght, product_photos_qty, product_weight_g, product_length_cm, product_height_cm, product_width_cm, order_id, order_item_id, seller_id, shipping_limit_date, price, freight_value]
Index: []


In [75]:
# Q1: Total Sellers
order_items["seller_id"].nunique()

3095

In [76]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [77]:
# Q2 — Seller-wise Orders.
order_sales = order_items.groupby("seller_id")["order_id"].count()

In [78]:
# Q3 — Top 10 sellers by number of items sold.
order_sales.sort_values(ascending = False).head(10)

seller_id
6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
955fee9216a65b617aa5c0531780ce60    1499
1025f0e2d44d7041d6cf58b6550e0bfa    1428
7c67e1448b00f6e969d365cea6b010ab    1364
ea8482cd71df3c1969d7b9473ff13abc    1203
7a67c85e85bb2ce8582c35f2203ad736    1171
Name: order_id, dtype: int64

In [79]:
# Q4 — Seller-wise Revenue
seller_revenue = order_items.groupby("seller_id")["price"].sum()

In [80]:
#  Q5 — Top 10 sellers by revenue.
seller_revenue.sort_values(ascending = False).head(10)

seller_id
4869f7a5dfa277a7dca6462dcf3b52b2    229472.63
53243585a1d6dc2643021fd1853d8905    222776.05
4a3ca9315b744ce9f8e9374361493884    200472.92
fa1c13f2614d7b5c4749cbc52fecda94    194042.03
7c67e1448b00f6e969d365cea6b010ab    187923.89
7e93a43ef30c4f03f38b393420bc753a    176431.87
da8622b14eb17ae2831f4ac5b9dab84a    160236.57
7a67c85e85bb2ce8582c35f2203ad736    141745.53
1025f0e2d44d7041d6cf58b6550e0bfa    138968.55
955fee9216a65b617aa5c0531780ce60    135171.70
Name: price, dtype: float64

In [81]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [82]:
# Q6 — Average Selling Price per Seller
avg_selling_price = order_items.groupby("seller_id")["price"].mean()

In [83]:
#  Q7 — Seller-wise Freight Revenue
seller_frieght = order_items.groupby("seller_id")["freight_value"].sum()

In [84]:
# Q8 — Seller Order Contribution %

seller_contribution = (order_sales / (order_sales.sum())) *100


In [85]:
#  Q9 — Sellers with Very Low Sales
order_sales.describe()
low_sales_sellers = order_sales[
    order_sales <= 2
]
low_sales_sellers.shape

(837,)

### ### Seller Analysis — Business Observations

1. **Total Sellers:** The dataset contains **3,095 unique sellers**, indicating a large and diverse seller base.

2. **Seller-wise Orders:** Seller order volumes vary significantly, with some sellers handling many more order-item records than others.

3. **Top 10 Sellers by Items Sold:** A small group of sellers contributes a large number of sold items, indicating that sales volume is not evenly distributed across all sellers.

4. **Seller-wise Revenue:** Revenue differs considerably between sellers because both the number of items sold and their selling prices vary.

5. **Top 10 Sellers by Revenue:** The highest-revenue sellers represent the sellers generating the greatest total sales value and are important contributors to overall marketplace revenue.

6. **Average Selling Price per Seller:** Average selling prices differ across sellers, showing that sellers operate across different product price ranges.

7. **Seller-wise Freight Revenue:** Freight value also varies between sellers, depending on the number of items sold and the freight charges associated with their orders.

8. **Seller Order Contribution:** Seller contribution percentages show how much each seller contributes to the overall order-item volume. This helps identify whether marketplace activity is concentrated among a smaller group of sellers.

9. **Very Low Sales Sellers:** The sales distribution shows a **median of only 8 order-item records per seller**, while the minimum is 1. This indicates that many sellers have relatively low sales volume.

10. **Overall Business Observation:** The seller base is large, but sales are **unevenly distributed**. A smaller group of sellers generates substantially higher sales volumes, while many sellers operate with relatively low order activity. This suggests that seller performance varies considerably across the marketplace.


## Customer Analysis

**Status:** Not Started

**Next Step:** Load `customers` dataset and begin Customer Analysis.

In [86]:
customers = pd.read_csv(
    r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_customers_dataset.csv"
    )
customers.shape

(99441, 5)

In [87]:
# Q1 — Total Unique Customers.
customers["customer_unique_id"].nunique()

96096

In [88]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')

In [89]:
unsold_check = products.merge(
    order_items,
    on="product_id",
    how="left"
)

In [90]:
order_items

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14
...,...,...,...,...,...,...,...
112645,fffc94f6ce00a00581880bf54a75a037,1,4aa6014eceb682077f9dc4bffebc05b0,b8bc237ba3788b23da09c0f1f3a3288c,2018-05-02 04:11:01,299.99,43.41
112646,fffcd46ef2263f404302a634eb57f7eb,1,32e07fd915822b0765e448c4dd74c828,f3c38ab652836d21de61fb8314b69182,2018-07-20 04:31:48,350.00,36.53
112647,fffce4705a9662cd70adb13d4a31832d,1,72a30483855e2eafc67aee5dc2560482,c3cfdc648177fdbbbb35635a37472c53,2017-10-30 17:14:25,99.90,16.95
112648,fffe18544ffabc95dfada21779c9644f,1,9c422a519119dcad7575db5af1ba540e,2b3e4a2a3ea8e01938cabda2a3e5cc79,2017-08-21 00:04:32,55.99,8.72


In [91]:
#  Q2 — Customer-wise Orders
customers_orders = orders.merge(
        customers,
        on= "customer_id"
    )

In [92]:
customers_orders.shape

(99441, 16)

In [93]:
top_customers = customers_orders.groupby("customer_unique_id")["order_id"].count().sort_values(ascending = False).head(10)

In [94]:
top_customers

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
6469f99c1f9dfae7733b25662e7f1782     7
ca77025e7201e3b30c44b472ff346268     7
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
12f5d6e1cbf93dafd9dcc19095df0b3d     6
63cfc61cee11cbe306bff5857d00bfe4     6
dc813062e0fc23409cd255f7f53c7074     6
de34b16117594161a6a89c50b289d35a     6
Name: order_id, dtype: int64

In [95]:
#  Q4 — Customer-wise Revenue
 
#   task 1 merge a customers_orders and order_items

customers_sales = customers_orders.merge(
    order_items,
    on = "order_id"
)

customer_revenue = customers_sales.groupby("customer_unique_id")["price"].sum()

In [96]:
customers_sales.shape

(112650, 22)

In [97]:
customers_sales.groupby("customer_unique_id")["price"].sum().sort_values(ascending=False).head(10)

customer_unique_id
0a0a92112bd4c708ca5fde585afaa872    13440.0
da122df9eeddfedc1dc1f5349a1a690c     7388.0
763c8b1c9c68a0229c42c9fc6f662b93     7160.0
dc4802a71eae9be1dd28f5d788ceb526     6735.0
459bef486812aa25204be022145caa62     6729.0
ff4159b92c40ebe40454e3e6a7c35ed6     6499.0
4007669dec559734d6f53e029e360987     5934.6
eebb5dda148d3893cdaf5b5ca3040ccb     4690.0
5d0a2980b292d049061542014e8960bf     4599.9
48e1ac109decbb87765a3eade6854098     4590.0
Name: price, dtype: float64

In [98]:
# Q6 — Average Customer Order Value (AOV)
customers_sales.groupby("customer_unique_id")["price"].sum().mean()

np.float64(142.44019807168308)

In [99]:
#  Q7 — One-time vs Repeat Customers
x = customers_orders.groupby("customer_unique_id")["order_id"].count()
y =x[x==1]
len(y)


z = x[x>1]
len(z)


2997

In [100]:
customers_sales.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'delivery_days', 'orders_months', 'Delayed_orders', 'delay_days',
       'customer_unique_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [101]:
#  Q8 — Customer Contribution %
(customer_revenue/ (customers_sales["price"].sum()))*100

customer_unique_id
0000366f3b9a7992bf8c76cfdf3221e2    0.000956
0000b849f77a49e4a4ce2b2a4ca5be3f    0.000139
0000f46a3911fa3c0805444483337064    0.000508
0000f6ccb0745a6a4b88665a16c9f078    0.000191
0004aac84e0df4da2b147fca70cf8255    0.001324
                                      ...   
fffcf5a5ff07b0908bd4e2dbc735a684    0.011551
fffea47cd6d3cc0a88bd621562a9d061    0.000477
ffff371b4d645b6ecea244b27531430a    0.000661
ffff5962728ec6157033ef9805bacc48    0.000846
ffffd2657e2aad2907e67c3e9daecbeb    0.000419
Name: price, Length: 95420, dtype: float64

In [102]:
#  Q9: Low-Value Customers
customer_revenue[customer_revenue <100]

customer_unique_id
0000b849f77a49e4a4ce2b2a4ca5be3f    18.90
0000f46a3911fa3c0805444483337064    69.00
0000f6ccb0745a6a4b88665a16c9f078    25.99
00050ab1314c0e55a6ca13cf7181fecf    27.99
0006fdc98a402fceb4eb0ee528f6a8d4    13.90
                                    ...  
fffa431dd3fcdefea4b1777d114144f2    64.99
fffb09418989a0dbff854a28163e47c6    58.00
fffea47cd6d3cc0a88bd621562a9d061    64.89
ffff371b4d645b6ecea244b27531430a    89.90
ffffd2657e2aad2907e67c3e9daecbeb    56.99
Name: price, Length: 54534, dtype: float64

### Customer Business Observations

- The majority of customers are one-time buyers, indicating that repeat purchase behavior is relatively limited.
- A smaller portion of customers are repeat buyers, showing an opportunity to improve customer retention and encourage repeat purchases.
- Customers generating less than ₹100 in revenue represent low-value customers and can be targeted with personalized offers, discounts, or product recommendations.
- Overall, customer retention and increasing repeat purchases can help improve long-term customer revenue.

In [103]:
reviews = pd.read_csv(
    r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_order_reviews_dataset.csv"
)

In [104]:
#  Q1 — Reviews dataset load karo
reviews.shape


(99224, 7)

In [105]:
reviews.columns


Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='str')

In [106]:
reviews.isnull().sum()


review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [107]:
reviews.dtypes


review_id                    str
order_id                     str
review_score               int64
review_comment_title         str
review_comment_message       str
review_creation_date         str
review_answer_timestamp      str
dtype: object

In [108]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 17.8 MB


In [109]:
# Q2 — Review Score ka distribution
reviews["review_score"].value_counts()

review_score
5    57328
4    19142
1    11424
3     8179
2     3151
Name: count, dtype: int64

In [110]:
#  Q3 — Average Review Score
reviews["review_score"].mean()

np.float64(4.08642062404257)

In [111]:
#  Q4 — Review Score Percentage Distribution
reviews["review_score"].value_counts(normalize= True)* 100

review_score
5    57.776344
4    19.291704
1    11.513344
3     8.242965
2     3.175643
Name: proportion, dtype: float64

In [112]:
#  Q5 — Review IDs aur Order IDs
reviews["review_id"].nunique()
98410
reviews["order_id"].nunique()
98673

98673

In [113]:
#  Q6. Actual duplicated rows
reviews.duplicated().sum()

np.int64(0)

In [114]:
#  Q7 — Repeated Review IDs
result = reviews["review_id"].value_counts()
result[result > 1]

review_id
c444278834184f72b1484dfe47de7f97    3
308316408775d1600dad81bd3184556d    3
2d6ac45f859465b5c185274a1c929637    3
3415c9f764e478409e8e0660ae816dd2    3
4219a80ab469e3fc9901437b73da3f75    3
                                   ..
8e954c79dc2fc426d5f17035e9bb22dd    2
b54d02f7f1520b3995f84b77ab0dacb3    2
870d856a4873d3a67252b0c51d79b950    2
9ea72755ef171e35b7b68ec0c8ed822c    2
d23bba9a2f1d16e5505a02e5968c1e68    2
Name: count, Length: 789, dtype: int64

In [115]:
# Q8 — Repeated Review ID ka actual pattern
reviews[reviews["review_id"] == "c444278834184f72b1484dfe47de7f97"]

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
2952,c444278834184f72b1484dfe47de7f97,df56136b8031ecd28e200bb18e6ddb2e,5,NaN,NaN,2017-02-08 00:00:00,2017-02-14 13:58:48
43897,c444278834184f72b1484dfe47de7f97,566f53fc7d36fa366f7e221468121877,5,NaN,NaN,2017-02-08 00:00:00,2017-02-14 13:58:48
95774,c444278834184f72b1484dfe47de7f97,8557dabbdacec1a9e250f5e70afe1eab,5,NaN,NaN,2017-02-08 00:00:00,2017-02-14 13:58:48


In [116]:
# Q9 — Order-level review uniqueness
reviews["order_id"].value_counts()[reviews["order_id"].value_counts() > 1]

order_id
c88b1d1b157a9999ce368f218a407141    3
df56136b8031ecd28e200bb18e6ddb2e    3
03c939fd7fd3b38f8485a0f95798f1f6    3
8e17072ec97ce29f0e1f111e598b0c85    3
cf73e2cb1f4a9480ed70c154da3d954a    2
                                   ..
e00ed9d20c3479f9f0e9727ca9d60946    2
5c2049caca936695cfa43805ebe7d8e3    2
6066fbcebeaf378770ad9fbe8e30d1f0    2
2739713764165279791b394484947e2d    2
157de1652c032429a625d034be79eeeb    2
Name: count, Length: 547, dtype: int64

In [117]:
# Q10 — Multiple reviews wale orders ka review score check
reviews[reviews["order_id"] == "c88b1d1b157a9999ce368f218a407141"]

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
1985,ffb8cff872a625632ac983eb1f88843c,c88b1d1b157a9999ce368f218a407141,3,NaN,NaN,2017-07-22 00:00:00,2017-07-26 13:41:07
82525,202b5f44d09cd3cfc0d6bd12f01b044c,c88b1d1b157a9999ce368f218a407141,5,NaN,NaN,2017-07-22 00:00:00,2017-07-26 13:40:22
89360,fb96ea2ef8cce1c888f4d45c8e22b793,c88b1d1b157a9999ce368f218a407141,5,NaN,NaN,2017-07-21 00:00:00,2017-07-26 13:45:15


In [118]:
r = reviews.groupby("order_id")["review_score"].nunique()
r[r > 1]

order_id
013056cfe49763c6f66bda03396c5ee3    2
02355020fd0a40a0d56df9f6ff060413    2
029863af4b968de1e5d6a82782e662f5    2
03c939fd7fd3b38f8485a0f95798f1f6    2
03eba6d9fef8f5b3e811d4b5a7cca9cd    2
                                   ..
fa350da9efbba83c2957b629bf6c2c7c    2
fc3d87913825143fa41c48e40267c47a    2
fcde2f7449493b4f207a23fae117b0fb    2
fd61441ba2a7b57e6342862e779b10b0    2
ff763b73e473d03c321bcd5a053316e8    2
Name: review_score, Length: 202, dtype: int64

In [119]:
# Q12 — ab final useful check
reviews.groupby("order_id")["review_score"].nunique().value_counts()

review_score
1    98471
2      202
Name: count, dtype: int64

In [120]:
#  Q13: Review comments ka coverage
reviews["review_comment_title"].notna().sum()

np.int64(11568)

In [121]:
#  Q14. Check the review send feedback
reviews["review_comment_message"].notna().sum()

np.int64(40977)

In [122]:
#  Q15. Reviews aur Orders ka relationship
reviews["order_id"].isin(orders["order_id"]).sum()

np.int64(99224)

In [123]:
orders["order_id"].isin(reviews["order_id"]).sum()

np.int64(98673)

### Reviews Dataset — Business Observations

* The reviews dataset contains **99,224 review records** across **7 columns**.
* The average review score is approximately **4.09 out of 5**, indicating that the overall recorded ratings are relatively positive.
* **5-star reviews are the most common**, followed by 4-star and 1-star reviews.
* There are **no exact duplicate rows** in the reviews dataset.
* The dataset contains **789 repeated `review_id` values**, indicating that some review IDs occur multiple times.
* **547 `order_id` values** appear more than once in the reviews dataset.
* Among these, **202 orders have multiple different review scores**, showing that some orders are associated with more than one distinct rating.
* All **99,224 review records** have a corresponding `order_id` in the orders dataset.
* Out of **99,441 orders**, **98,673 orders** have at least one associated review record, meaning **768 orders do not have a review record**.
* Written review messages are missing for **58,247 reviews**, while **40,977 reviews contain a review message**.
* Missing review titles and messages should **not be automatically removed during exploration**; their treatment will be decided during the Data Cleaning phase based on business requirements.


In [124]:
# Q16 — Review dates ka range
reviews["review_creation_date"].min()

'2016-10-02 00:00:00'

In [125]:
reviews["review_creation_date"].max()

'2018-08-31 00:00:00'

In [126]:
# Next Q17
reviews['review_answer_timestamp'].min()

'2016-10-07 18:32:28'

In [127]:
reviews["review_answer_timestamp"].max()

'2018-10-29 12:27:35'

In [128]:
# Q18 — ab proper datetime conversion
reviews["review_creation_date"] = pd.to_datetime(reviews["review_creation_date"])
reviews["review_answer_timestamp"] = pd.to_datetime(reviews["review_answer_timestamp"])

In [129]:
reviews.dtypes

review_id                             str
order_id                              str
review_score                        int64
review_comment_title                  str
review_comment_message                str
review_creation_date       datetime64[us]
review_answer_timestamp    datetime64[us]
dtype: object

In [130]:
# Q19 — Response time
reviews["response_time"] = (
    reviews["review_answer_timestamp"] -
    reviews["review_creation_date"]
)
reviews["response_time"].describe()

count                     99224
mean     3 days 03:34:33.029700
std      9 days 21:21:40.258026
min             0 days 02:08:29
25%      1 days 00:07:00.750000
50%      1 days 16:11:55.500000
75%             3 days 02:29:08
max           518 days 16:46:52
Name: response_time, dtype: object

In [131]:
# Q20 — response time ko days mein convert karo
reviews["response_days"] = reviews["response_time"].dt.total_seconds() / (24 * 60 * 60)

In [132]:
reviews["response_days"].describe()

count    99224.000000
mean         3.148993
std          9.890049
min          0.089225
25%          1.004870
50%          1.674948
75%          3.103565
max        518.699213
Name: response_days, dtype: float64

### Reviews Dataset — Final Business Observations

* The reviews dataset contains **99,224 review records** across **7 columns**.
* The average review score is approximately **4.09 out of 5**, with 5-star reviews being the most common.
* There are **no exact duplicate rows** in the dataset.
* **789 `review_id` values** appear more than once, while **547 `order_id` values** have multiple review records.
* **202 orders** have multiple review records with different review scores.
* All **99,224 review records** are linked to valid orders in the orders dataset.
* Out of **99,441 orders**, **98,673 orders** have an associated review record; **768 orders** do not have a review record.
* **40,977 reviews** contain a written review message, while **58,247 reviews** have no review message.
* Review creation dates range from **October 2016 to August 2018**.
* Review answer timestamps range from **October 2016 to October 2018**.
* The average review response time is approximately **3.15 days**, while the median is approximately **1.67 days**.
* The maximum recorded response time is approximately **518.70 days**, indicating some extreme long-response records that should be investigated during the Data Quality/Cleaning phase.
* Missing review titles and messages will be handled later according to business requirements rather than being automatically removed during exploration.


# Location Data Set load 

In [133]:
geolocation = pd.read_csv(
    r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\olist_geolocation_dataset.csv"
)

In [134]:
#  Q1 
geolocation.shape

(1000163, 5)

In [135]:
#  Q2. 
geolocation.columns

Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='str')

In [136]:
# Q3. check the datatypes
geolocation.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object

In [137]:
# Q4 — Null Values
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [138]:
# Q5 — Exact Duplicates.
geolocation.duplicated().sum()

np.int64(261831)

In [139]:
#  Q6
geolocation["geolocation_zip_code_prefix"].nunique()

19015

In [140]:
# Q7.
geolocation["geolocation_city"].nunique()

8011

In [143]:
# Q8 — Unique States
geolocation["geolocation_state"].nunique()

27

In [147]:
# Q9 — Latitude/Longitude.
geolocation[["geolocation_lat", "geolocation_lng"]].describe()

,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06
mean,-2.117615e+01,-4.639054e+01
std,5.715866e+00,4.269748e+00
min,-3.660537e+01,-1.014668e+02
25%,-2.360355e+01,-4.857317e+01
50%,-2.291938e+01,-4.663788e+01
75%,-1.997962e+01,-4.376771e+01
max,4.506593e+01,1.211054e+02


In [149]:
#  Q10 — ZIP Prefix vs City/State.
geolocation.groupby("geolocation_zip_code_prefix")[["geolocation_city", "geolocation_state"]].nunique()

,geolocation_city,geolocation_state
geolocation_zip_code_prefix,,
1001,2,1
1002,2,1
1003,2,1
1004,2,1
1005,2,1
...,...,...
99960,1,1
99965,1,1
99970,2,1


In [150]:
#  Q11 — Repeated ZIP Prefixes.
geolocation["geolocation_zip_code_prefix"].value_counts().describe()

count    19015.000000
mean        52.598633
std         72.057907
min          1.000000
25%         10.000000
50%         29.000000
75%         66.500000
max       1146.000000
Name: count, dtype: float64

### Final Business Observations

- Dataset mein 1,000,163 records aur 5 columns hain.
- Dataset mein 19,015 unique ZIP prefixes, 8,011 unique cities aur 27 states - represented hain.
- Kisi bhi column mein missing values nahi hain, isliye geographic information complete hai.
- 261,831 exact duplicate rows mile, jo data-quality investigation ke liye important observation hai.
- Latitude aur longitude numeric format mein hain, lekin kuch unusual coordinate values bhi present hain, jinhe Phase 2 mein validate karna chahiye.
- ZIP prefix ke andar multiple city names mil sakte hain, jabki observed mapping mein ZIP prefix ek state ke andar hi represented hai.
- ZIP prefixes ki record frequency uneven hai: average 52.6 records, median 29, aur maximum 1,146 records per ZIP prefix.
- Overall, dataset customer/seller locations ko ZIP → City → State → Coordinates level par analyze karne ke liye useful geographic information provide karta hai.

### product_category_name_translation dataset.

In [ ]:
category_translation = pd.read_csv(r"C:\Users\MAINAK\Desktop\Data-Science\projects\MarketLens\data\raw\product_category_name_translation.csv")
#  Q1. columns shape
category_translation.shape

(71, 2)

In [ ]:
#  Q2. coloumns classification 
category_translation.columns

Index(['product_category_name', 'product_category_name_english'], dtype='str')

In [ ]:
#  Q3. dtypes 
category_translation.dtypes

product_category_name            str
product_category_name_english    str
dtype: object

In [ ]:
#  Q4. null values
category_translation.notnull().sum()

product_category_name            71
product_category_name_english    71
dtype: int64

In [157]:
# Ab Q5 — Exact Duplicates
category_translation.duplicated().sum()

np.int64(0)

In [160]:
#  Q6 — Unique Categories
category_translation["product_category_name"].nunique()

# category_translation["product_category_name"].unique()

71

In [161]:
#  Q7 — English categories
category_translation["product_category_name_english"].nunique()

71

In [162]:
# Q8 — Mapping consistency
category_translation.groupby("product_category_name")["product_category_name_english"].nunique()

product_category_name
agro_industria_e_comercio    1
alimentos                    1
alimentos_bebidas            1
artes                        1
artes_e_artesanato           1
                            ..
sinalizacao_e_seguranca      1
tablets_impressao_imagem     1
telefonia                    1
telefonia_fixa               1
utilidades_domesticas        1
Name: product_category_name_english, Length: 71, dtype: int64

### Final Business Observations

- Dataset contains 71 product category mappings.
- No missing values were found in either column.
- No exact duplicate rows were found.
- All 71 original product categories have corresponding English translations.
- Each original category maps to exactly one English category, showing consistent 1-to-1 mapping.
- This dataset can be used to translate Portuguese product categories into English for clearer business analysis and reporting.